tools/azure-devtools/src/azure_devtools/scenario_tests/preparers.py

```diff
@@ -59,18 +112,47 @@ def _prepare_create_resource(self, test_class_instance, **kwargs):
 
     def __call__(self, fn):
         def _preparer_wrapper(test_class_instance, **kwargs):
+            _logger.debug("Entering preparer wrapper for %s and test %s",
+                self.__class__.__name__, str(test_class_instance))
 
-            resource_name, kwargs = self._prepare_create_resource(test_class_instance, **kwargs)
+            resource_name, kwargs = self._prepare_create_resource(test_class_instance, fn, **kwargs)
 
             trim_kwargs_from_test_function(fn, kwargs)
 
             try:
+                test_class_instance._delay_removal = False # pylint: disable=protected-access
                 fn(test_class_instance, **kwargs)
-            finally:
-                # Russian Doll - the last declared resource to be deleted first.
-                self.remove_resource_with_record_override(resource_name, **kwargs)
-
+                if self._use_cache: # If we're cached, stop things above us (e.g. RG) from deleting.
+                    test_class_instance._delay_removal = True # pylint: disable=protected-access
+            finally:              
+                def _deferred_delete():
+                    _logger.debug("Running deferred delete for %s", self)
+                    self.remove_resource_with_record_override(resource_name, **kwargs)
+                # If we use cache we delay deletion for the end.
+                # This won't guarantee deletion order, but it will guarantee everything delayed
+                # does get deleted, in the worst case by getting rid of the RG at the top.
+                if self._use_cache:
+                    # Only the first occurance though; since they point to the same underlying, and the
+                    # latter cache-fetched-instances didn't actually initialize ARM clients to be able to delete.
+                    if self._aggregate_cache_key not in AbstractPreparer._cache_pending_deletes:
+                        with self._cache_lock:
+                            AbstractPreparer._cache_pending_deletes[self._aggregate_cache_key] = _deferred_delete
+                elif test_class_instance._delay_removal: # pylint: disable=protected-access
+                    # If a child is specifying a cache but we aren't, raise an exception.
+                    # If it were ever desired to do implied caching, one would simply add a deferred delete here.
+                    # raise Exception("Preparer Error: If a child resource is cached, parents must also be cached")
```

Comment:
Note to self:  This comment can be removed and replaced with a comment denoting that this section is relevant if the check on line 58 ever goes away.